sympy/core/traversal.py

```diff
@@ -5,6 +5,40 @@
 
 
 
+def apply(f, e, atoms=True, nonbasic=True):
```

Comment:
I don't see the purpose of the `nonbasic` argument. At least there should be a better name for whatever it does.

Reply:
It has the same meaning as for `bottom_up` (where it was originally used); it controls whether a nonbasic arg is to be affected or not.

Reply:
Why not just apply it anyway to nonbasic?

My guess is that the nonbasic argument was added to workaround having:

1. Functions that did not work with unsympified types.
2. Buggy types that have non-Basic in args.

The better solution in both cases is either that the function should accept unsympified types or that it should not be given unsympified types and that any buggy type with non-Basic in args should be fixed.

I also don't really see the point of the atoms parameter. It would be better if `apply` had a simple syntax that is just like `map` except that it works for an arbitrary nested list/tuple of things.

Checking with git grep suggests that the nonbasic and atoms parameters for `bottom_up` are not used anywhere (not even in `bottom_up` tests).

Reply:
If someone wants to avoid certain arguments, those rules could be part of the traversal function.